### PR TITLE
[3/n][dagster-aws] Port S3Resource, S3IOManager to Pythonic resources

### PR DESCRIPTION
**Pull request by benpankow**
## Summary

Finishes off the `dagster-aws` migration by migrating the `S3IOManager` and `S3Resource` over.

## Test Plan

Add new resources to existing unit tests.

### CODE REVIEW

**Comment by benpankow** [comment]
Current dependencies on/for this PR:
* master
  * **PR #13430** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/13430" target="_blank">Graphite</a> 
    * **PR #13419** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/13419" target="_blank">Graphite</a> 
      * **PR #13421** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/13421" target="_blank">Graphite</a> 
        * **PR #13433** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/13433" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/13433?utm_source=stack-comment).

File: python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py
Comment (schrockn) [inline python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py:151]: update docblock ot `Definitions`, etc

File: python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py
Comment (schrockn) [inline python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py:153]: `ResourceDependency[S3Resource]` ?

File: python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py
Comment (schrockn) [inline python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py:164]: seems like we have another naming problem here in terms of `S3PickleIOManager` versus `PickledObjectS3IOManager`

File: python_modules/libraries/dagster-aws/dagster_aws/s3/resources.py
Comment (schrockn) [inline python_modules/libraries/dagster-aws/dagster_aws/s3/resources.py:16]: same discussion to be had as the upstream diff about these "resource base" classes

File: python_modules/libraries/dagster-aws/dagster_aws/s3/resources.py
Comment (schrockn) [inline python_modules/libraries/dagster-aws/dagster_aws/s3/resources.py:83]: update docblock

File: python_modules/libraries/dagster-aws/dagster_aws/s3/resources.py
Comment (schrockn) [inline python_modules/libraries/dagster-aws/dagster_aws/s3/resources.py:182]: the more I see `create_resource` the more I cringe here.

I think we should consider `create_object_for_user_code` or something which will also solve the same io_manager problem and will allow us to consolidate those two methods in the base class

**Review by schrockn — DISMISSED** [review]

File: python_modules/libraries/dagster-aws/dagster_aws/s3/resources.py
Comment (benpankow) [reply python_modules/libraries/dagster-aws/dagster_aws/s3/resources.py:16]: Updated to `ResourceWithS3Configuration`, for now

File: python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py
Comment (benpankow) [reply python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py:164]: Yeah, this is tricky since we need to maintain backcompat. This might be worth bikeshedding on a call.

File: python_modules/libraries/dagster-aws/dagster_aws/s3/resources.py
Comment (benpankow) [reply python_modules/libraries/dagster-aws/dagster_aws/s3/resources.py:183]: I can put up a PR for this - are we OK breaking the `create_resource` path for existing resources (e.g. users who are trying the experimental APIs out?). 

We could also do some magic to find an impl of `create_resource` and defer to that if `create_object_for_user_code` isn't overridden.

Comment (schrockn) [reply python_modules/libraries/dagster-aws/dagster_aws/s3/resources.py:183]: Yes it is experimental so that we have the latitude to do exactly that.

File: python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py
Comment (schrockn) [inline python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py:106]: i/o is just the abc

```
class S3IOManagerResource(FactoryResource[PickledObjectS3IOManager]):
  pass
```

according to our discussed naming convention

**Review by schrockn — CHANGES_REQUESTED** [review]
Popping back in your queue. Do rename as discusses and restack on new proposed naming scheme.

File: python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py
Comment (schrockn) [inline python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py:106]: even thought it is more verbose, I do think that it is more consistent and clear to call this `PickledObjectS3IOManagerResource`

**Review by schrockn — CHANGES_REQUESTED** [review]
makes me a bit sad, but I think we use the long name OR rename PickledObjectS3IOManager and have a consistent naming convention

File: python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py
Comment (schrockn) [inline python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py:107]: ConfigurablePickledObjectS3IOManager?

File: python_modules/libraries/dagster-aws/dagster_aws/s3/ops.py
Comment (schrockn) [inline python_modules/libraries/dagster-aws/dagster_aws/s3/ops.py:84]: I think we should hold off on merging this until https://github.com/dagster-io/dagster/pull/13495 is merged so we can demonstrate this as an example

File: python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py
Comment (schrockn) [inline python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py:160]: `inner_io_manager`

**Review by schrockn — CHANGES_REQUESTED** [review]
Some naming stuff.

File: python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py
Comment (schrockn) [inline python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py:173]: @sryza this is still definitely making me wince. I think this will be copied and pasted throughout the land and it's quite a bit of boilerplate, in the end it's extremely similar behavior.

Definitely still think we should support people making IOManagers directly configurable but forcing us to do these delegation shenanigans rather than just support factory directly seems like a lot.

**Review by schrockn — CHANGES_REQUESTED** [review]
Def should not name it PickledObjectS3IOManagerResource

File: python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py
Comment (schrockn) [reply python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py:173]: Regardless of the wisdom of invoking list_objects as a health check, its there and removing it is a large behavior change that will potentially break users if we remove it.

File: python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py
Comment (schrockn) [inline python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py:163]: wait what is the behavior here? I think this will not work? we need to call `self.s3_resource.create_client()` no?

**Review by schrockn — CHANGES_REQUESTED** [review]
much better with `IAttachDifferentObjectToOpContext`. However I'm not convinced that that resource dependency works as advertised.

File: python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py
Comment (benpankow) [reply python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py:163]: You're right. The test doesn't catch this since it was passing in an old-style S3 resource. Updated tests + this code accordingly.

**Review by schrockn — APPROVED** [review]

**Pull request merged**
